src/common/storage/src/stage.rs

```diff
@@ -29,3 +50,185 @@ pub fn init_stage_operator(stage_info: &UserStageInfo) -> Result<Operator> {
         Ok(pop.layer(SubdirLayer::new(&stage_info.stage_prefix())))
     }
 }
+/// select * from @s1/<path> (FILES => <files> PATTERN => <pattern>)
+/// copy from @s1/<path> FILES = <files> PATTERN => <pattern>
+#[derive(serde::Serialize, serde::Deserialize, Clone, PartialEq, Eq, Debug)]
+pub struct StageFilesInfo {
+    pub path: String,
+    pub files: Option<Vec<String>>,
+    pub pattern: Option<String>,
+}
+
+impl StageFilesInfo {
+    fn get_pattern(&self) -> Result<Option<Regex>> {
+        match &self.pattern {
+            Some(pattern) => match Regex::new(pattern) {
+                Ok(r) => Ok(Some(r)),
+                Err(e) => Err(ErrorCode::SyntaxException(format!(
+                    "Pattern format invalid, got:{}, error:{:?}",
+                    pattern, e
+                ))),
+            },
+            None => Ok(None),
+        }
+    }
+
+    pub async fn list(&self, operator: &Operator, first_only: bool) -> Result<Vec<FileWithMeta>> {
+        if let Some(files) = &self.files {
+            let mut res = Vec::new();
+            for file in files {
+                let full_path = Path::new(&self.path)
+                    .join(file)
+                    .to_string_lossy()
+                    .to_string();
+                let meta = operator.object(&full_path).metadata().await?;
+                if meta.mode().is_file() {
+                    res.push(FileWithMeta::new(&full_path, meta))
+                } else {
+                    return Err(ErrorCode::BadArguments(format!(
+                        "{full_path} is not a file"
+                    )));
+                }
+                if first_only {
+                    break;
+                }
+            }
+            Ok(res)
+        } else {
+            let pattern = self.get_pattern()?;
+            list_files_with_pattern(operator, &self.path, pattern, first_only).await
+        }
+    }
+
+    pub async fn first_file(&self, operator: &Operator) -> Result<FileWithMeta> {
+        let mut files = self.list(operator, true).await?;
```

Comment:
If we should list all files at once and hold in memory? Since network IO is expensive.

Reply:
here we list only one file (one list_dir op for local fs )
I think a pro of listing all file in read_partition (instead of binding->add_table) is we can report some syntax error earlier if read_partition take a long time.

is there other diff?

cc @BohuTANG  @Xuanwo  @sundy-li

Reply:
What's the read path of parquet in this patch?

Reply:
read the first file when binding to infer schema,
read all files in read_partition.

Reply:
I think we should always do list since the result will change.